tests/tests.cpp

```diff
@@ -0,0 +1,398 @@
+
+        // 4) Decrypt with the wrong key
+        uint8_t new_key[CIPHER_KEY_SIZE];
+        crypto.RandGen(new_key, CIPHER_KEY_SIZE);
+
+        ASSERT_NE(0, crypto.SymDec(new_key, ct.data(),
+                                   reinterpret_cast<const uint8_t*>("aad"),
+                                   pt.data(), ct.size(), strlen("aad")))
+            << "Ciphertext decrypted successfully with incorrect key";
+
+        // 5) Remove last block of ciphertext
+        perturbed_ct = std::vector<uint8_t>(ct);
+        perturbed_ct.resize(perturbed_ct.size() - CIPHER_KEY_SIZE);
+
+        ASSERT_NE(0,
+                  crypto.SymDec(sym_key, perturbed_ct.data(),
+                                reinterpret_cast<const uint8_t*>("aad"),
+                                pt.data(), perturbed_ct.size(), strlen("aad")))
+            << "Truncated ciphertext decrypted successfully";
+
+        // 6) Swap random blocks in the ciphertext
+        for (int i = 0; i < 100; i++) {
+            perturbed_ct = std::vector<uint8_t>(ct);
+            std::vector<uint8_t> ct_blocks(perturbed_ct.begin() +
+                                               CIPHER_IV_SIZE + CIPHER_TAG_SIZE,
+                                           perturbed_ct.end());
+            swap_blocks(ct_blocks, CIPHER_KEY_SIZE);
+            for (int j = CIPHER_IV_SIZE + CIPHER_TAG_SIZE;
+                 j < perturbed_ct.size(); j++) {
+                perturbed_ct[j] = ct_blocks[j];
+            }
+
+            ASSERT_NE(0, crypto.SymDec(sym_key, perturbed_ct.data(),
+                                       reinterpret_cast<const uint8_t*>("aad"),
+                                       pt.data(), perturbed_ct.size(),
+                                       strlen("aad")))
+                << "Ciphertext decrypted successfully with swapped blocks";
+        }
+    }
+
+    TEST_F(SoundnessTest, AsymEnc) {
+        // Encrypt the message
+        std::vector<uint8_t> ct(crypto.AsymEncSize(msg.size()));
+        ASSERT_EQ(0,
+                  crypto.AsymEnc(pem_key, msg.data(), ct.data(), msg.size()));
+
+        // We don't care about the value of the decryptions (they all should
+        // fail) so we can use the same plaintext vector for each decryption
+        std::vector<uint8_t> pt(crypto.AsymDecSize(ct.size()));
+        size_t _pt_size;
+
+        // 1) Make random pertubations in the ciphertexts
+        for (int i = 0; i < 100; i++) {
+            std::vector<uint8_t> perturbed_ct(ct);
+            perturb(perturbed_ct);
+            ASSERT_NE(0, crypto.AsymDec(perturbed_ct.data(), pt.data(),
+                                        perturbed_ct.size(), &_pt_size))
+                << "Perturbed ciphertext decrypted successfully";
+        }
+
+        // 2) Decrypt with the wrong key
+        // Load public key from file
+        std::vector<uint8_t> new_key;
+        load_file(std::string("../../tests/data/public.pub"), new_key);
```

Comment:
what key was used to encrypt originally? a randomly generated public key?

Reply:
Yes, each test generates a `UtilsTest` instance which randomly generates a new public key

Reply:
can you add a note on line 303 saying that `crypto.AsymDec()` uses the `UtilsTest` generated key? I think for someone that doesn't know the codebase very well it may be a bit confusing if we don't clarify this